File: mcp-code-intelligence-python/src/mcp_code_intel/memory/embedding/service.py
"""Service layer — coordinates embedding generation and storage."""

import math
import struct
import sys

from .provider import EmbeddingProvider


class EmbeddingService:
    """Wraps provider + vector repo for embed-and-store workflow."""

    def __init__(self, provider: EmbeddingProvider, vector_repo) -> None:
        self._provider = provider
        self._vector_repo = vector_repo

    def embed_and_store(self, entry_id: int, text: str) -> bool:
        """Generate and store embedding for a knowledge entry."""
        vector = self._provider.embed(text)
        if vector is None:
            return False
        blob = self._float_list_to_bytes(vector)
        self._vector_repo.upsert(
            entry_id, blob, self._provider.model_name, self._provider.dimensions
        )
        return True

    def embed_batch_and_store(self, entries: list[tuple[int, str]]) -> int:
        """Embed multiple entries. Returns count of successes."""
        count = 0
        for entry_id, text in entries:
            if self.embed_and_store(entry_id, text):
                count += 1
        return count

    def embed(self, text: str) -> list[float] | None:
        """Get raw embedding for text (without storing)."""
        return self._provider.embed(text)

    def is_available(self) -> bool:
        """Check if embedding provider is available."""
        return self._provider.is_available()

    def close(self) -> None:
        """Release resources."""
        self._provider.close()

    @staticmethod
    def _float_list_to_bytes(arr: list[float]) -> bytes:
        """Convert float list to little-endian bytes."""
        return struct.pack(f"<{len(arr)}f", *arr)

    @staticmethod
    def _bytes_to_float_list(data: bytes) -> list[float]:
        """Convert little-endian bytes back to float list."""
        count = len(data) // 4
        return list(struct.unpack(f"<{count}f", data))

    @staticmethod
    def cosine_similarity(a: list[float], b: list[float]) -> float:
        """Cosine similarity between two vectors."""
        if len(a) != len(b):
            return 0.0
        dot = sum(x * y for x, y in zip(a, b))
        norm_a = math.sqrt(sum(x * x for x in a))
        norm_b = math.sqrt(sum(x * x for x in b))
        denom = norm_a * norm_b
        return dot / denom if denom > 0 else 0.0


def _log(msg: str) -> None:
    print(f"[embed-svc] {msg}", file=sys.stderr, flush=True)

File: backend/mcp-code-intelligence-nodejs/src/memory/embedding/service.ts
/**
 * EmbeddingService — coordinates embedding generation and storage.
 * Wraps provider + vector repo for embed-and-store workflow.
 */

import { EmbeddingProvider } from './provider.js';
import { VectorRepository } from '../vector-repo.js';

export class EmbeddingService {
  private readonly provider: EmbeddingProvider;
  private readonly vectorRepo: VectorRepository;

  constructor(provider: EmbeddingProvider, vectorRepo: VectorRepository) {
    this.provider = provider;
    this.vectorRepo = vectorRepo;
  }

  /** Generate and store embedding for a knowledge entry. */
  async embedAndStore(entryId: number, text: string): Promise<boolean> {
    const vector = await this.provider.embed(text);
    if (!vector) return false;
    const blob = floatListToBytes(vector);
    this.vectorRepo.upsert(entryId, blob, this.provider.modelName, this.provider.dimensions);
    return true;
  }

  /** Embed multiple entries. Returns count of successes. */
  async embedBatchAndStore(entries: Array<[number, string]>): Promise<number> {
    let count = 0;
    for (const [entryId, text] of entries) {
      if (await this.embedAndStore(entryId, text)) count++;
    }
    return count;
  }

  /** Get raw embedding for text (without storing). */
  async embed(text: string): Promise<number[] | null> {
    return this.provider.embed(text);
  }

  /** Check if embedding provider is available. */
  isAvailable(): boolean {
    return this.provider.isAvailable();
  }

  /** Release resources. */
  close(): void {
    this.provider.close();
  }

  /** Cosine similarity between two vectors. */
  static cosineSimilarity(a: number[], b: number[]): number {
    if (a.length !== b.length) return 0.0;
    let dot = 0, normA = 0, normB = 0;
    for (let i = 0; i < a.length; i++) {
      dot += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }
    const denom = Math.sqrt(normA) * Math.sqrt(normB);
    return denom > 0 ? dot / denom : 0.0;
  }
}

/** Convert float array to little-endian Buffer (Float32). */
export function floatListToBytes(arr: number[]): Buffer {
  const buf = Buffer.alloc(arr.length * 4);
  for (let i = 0; i < arr.length; i++) {
    buf.writeFloatLE(arr[i], i * 4);
  }
  return buf;
}

/** Convert little-endian Buffer back to float array. */
export function bytesToFloatList(data: Buffer): number[] {
  const count = data.length / 4;
  const result: number[] = [];
  for (let i = 0; i < count; i++) {
    result.push(data.readFloatLE(i * 4));
  }
  return result;
}
